### ISSUE TITLE
RFC: Localization of pathnames and streamlining of navigation APIs

### ISSUE THREAD

**Issue by amannn** [amannn/next-intl#402]
## Localization of pathnames

**Update:** This is available as of `next-intl@3.0.0-beta.13`, see [the announcement thread](https://twitter.com/jamannnnnn/status/1695096032072208746).

We currently provide minimal documentation for [localizing pathnames]( https://next-intl-docs.vercel.app/docs/routing/middleware#localizing-pathnames) (see also [the named routes example](https://github.com/amannn/next-intl/tree/feat/next-13-rsc/examples/example-next-13-named-routes)).

It would be helpful to have built-in support for this.

We could achieve this by accepting a route map:

```tsx
// The `pathnames` object holds pairs of internal
// and external paths. Based on the locale, the
// corresponding external path will be picked.
const pathnames = {
  // If all locales use the same pathname, a single
  // external path can be used for all locales.
  '/': '/',
  '/blog': '/blog',
 
  // If locales use different paths, you can
  // specify each external path per locale.
  '/about': {
    en: '/about',
    de: '/ueber-uns'
  },
 
  // Dynamic params are supported via square brackets
  '/news/[articleSlug]-[articleId]': {
    en: '/news/[articleSlug]-[articleId]',
    de: '/neuigkeiten/[articleSlug]-[articleId]'
  }
};
```

**Proposed docs:**
 - [Middleware: Localizing pathnames](https://next-intl-docs-git-feat-next-13-rsc-pathname-l-93913f-next-intl.vercel.app/docs/routing/middleware#localizing-pathnames)
 - [New navigation docs](https://next-intl-docs-git-feat-next-13-rsc-pathname-l-93913f-next-intl.vercel.app/docs/routing/navigation)

**Prior art:** https://v8.i18n.nuxtjs.org/guide/custom-paths

## Streamlining of navigation APIs

Similar to the proposed factory function `createLocalizedPathnamesNavigation`, we could offer a `createSharedPathnamesNavigation` factory function that can be used to create the currently existing navigation APIs for non-localized pathnames.

This would:

1. Streamline all imports to a single namespace
2. Provide strict typing for the locales
3. Make it easier to upgrade from shared pathnames to localized pathnames (you'll have to adjust pathnames with dynamic params though, but we can provide TypeScript support here)

```tsx
import {createSharedPathnamesNavigation} from 'next-intl/navigation';
 
export const locales = ['en', 'de'] as const;
 
export const {Link, redirect, usePathname, useRouter} =
  createSharedPathnamesNavigation({locales});
```

Potentially we could furthermore return a typed version of `useLocale` from these functions.

**Rollout plan**: Introduce `createSharedPathnamesNavigation` in a minor version, deprecate the previous imports and remove them in a major release.

**Comment by amannn** [comment]
See also https://github.com/vercel/next.js/issues/49761

**Comment by amannn** [comment]
I've started working on a proof-of-concept for this!

**Comment by amannn** [comment]
**Update:** This is available as of `next-intl@3.0.0-beta.12`, see [the announcement thread](https://twitter.com/jamannnnnn/status/1695096032072208746).

**Comment by TommySorensen** [comment]
@amannn This really helps me removing my own version of this in my middleware 🎉
Could the `pathnames` be async so i potentially could fetch the list of url from a CMS one pr. minute?

**Comment by amannn** [comment]
@TommySorensen Great to hear! 🙌

The pathnames for the middleware can be fetched async if you [compose the middleware](https://next-intl-docs.vercel.app/docs/routing/middleware#composing-other-middlewares).

As for the corresponding navigation APIs, you might not have a benefit there, because the point of them is really the static typing (which doesn't apply to your use case). How do you solve this currently? Do you provide wrappers for navigation APIs yourself that are based on CMS data?

**Comment by amannn** [comment]
This has been fully implemented in the 3.0 RC (see [announcement](https://next-intl-docs.vercel.app/blog/next-intl-3-0))

**Comment by HarleySalas** [comment]
I'd imagine many people would use resource id's as their pathname key, as opposed to the default locale pathname. When getting dynamic data, a lot of cms api's will not let you query for a pathname that is different than the requested locale. Since it seems that rewrites are used, the param passed to a page component will only ever be the pathname's key and it fails to provide the localized pathname to the page params. When using default locale, it will often times ignore the localized pathname value and just display the pathname's object key. It seems like an optimistic assumption and in my opinion, would be much better to always prioritize the current locale, if available, rather than the pathname key.

tldr: I believe the pathname's key should *never* be shown in the url bar, if a localized version exists within the object.

**Comment by amannn** [comment]
@HarleySalas Do you think you could include some concrete examples that demonstrate your use case? It's a bit hard for me to follow since the terminology is quite broad.
